### PR TITLE
EPMRPP-79186 || Modal window "Create Defect"/"Edit Defect" isn't closed when color code was changed than previous value was pasted

### PR DESCRIPTION
**Pull request by Vadim73i**

### CODE REVIEW

**Review by Bam6ycha — CHANGES_REQUESTED** [review]

File: app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx [hunk 1]
```diff
@@ -110,14 +110,18 @@ const AddEditDefectTypeModal = ({
...
   change,
+  currentValue,
+  initialValue,
 }) => {
   const { formatMessage } = useIntl();
   const dispatch = useDispatch();
 
   const [color, setColor] = useState();
 
+  const isFormChanged =
```
[the remark this answers is not in the document]
Comment (Vadim73i) [reply app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:122]: As you can see, `dirty` has been used before and there is a bug.
As far as I know, the form should close if there were no changes.

[the remark this answers is not in the document]
Comment (Bam6ycha) [reply app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:122]: Okay, I see.

File: app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx [hunk 2]
```diff
@@ -243,27 +245,35 @@ AddEditDefectTypeModal.propTypes = {
...
-        shortName: composeBoundValidators([
-          bindMessageToValidator(validate.required, 'shortRequiredFieldHint'),
-          bindMessageToValidator(validate.defectTypeShortName, 'defectShortNameHint'),
-        ])(shortName),
-      };
-    },
-  })(AddEditDefectTypeModal),
+  connect((state) => ({
+    currentValue: getFormValues('DefectTypeForm')(state),
+    initialValue: getFormInitialValues('DefectTypeForm')(state),
```
Comment (Bam6ycha) [inline app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:261]: defectType prop is the initial value of the form. You don't need this value.

File: app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx [hunk 3]
```diff
@@ -243,27 +245,35 @@ AddEditDefectTypeModal.propTypes = {
...
-        ])(longName),
-        shortName: composeBoundValidators([
-          bindMessageToValidator(validate.required, 'shortRequiredFieldHint'),
-          bindMessageToValidator(validate.defectTypeShortName, 'defectShortNameHint'),
-        ])(shortName),
-      };
-    },
-  })(AddEditDefectTypeModal),
+  connect((state) => ({
+    currentValue: getFormValues('DefectTypeForm')(state),
```
Comment (Bam6ycha) [inline app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:260]: My suggestion is to rewrite it using formValueSelector from `redux-form` and useSelector hooks.
And we can select only the value of the color field and compare it with defeType.color value.

**Review by Bam6ycha — CHANGES_REQUESTED** [review]

File: app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx [hunk 4]
```diff
@@ -110,14 +110,18 @@ const AddEditDefectTypeModal = ({
...
   initialize,
-  dirty,
   change,
+  currentValue,
+  initialValue,
 }) => {
   const { formatMessage } = useIntl();
   const dispatch = useDispatch();
 
   const [color, setColor] = useState();
```
Comment (Bam6ycha) [inline app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:120]: My suggestion is to rewrite this line to `  const [color, setColor] = useState(defectType.color);`

File: app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx [hunk 5]
```diff
@@ -177,7 +181,7 @@ const AddEditDefectTypeModal = ({
       okButton={okButton}
       cancelButton={cancelButton}
       onClose={() => dispatch(hideModalAction())}
-      allowCloseOutside={!dirty}
+      allowCloseOutside={!isFormChanged}
```
Comment (Bam6ycha) [inline app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:184]: And this to `color?.toLowerCase() === defectType.color`

**Review by Bam6ycha — CHANGES_REQUESTED** [review]

File: app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx (hunk 5 above)
Comment (Bam6ycha) [reply app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:183]: Okay, maybe I didn't understand something.

But anyway, looks like now we can combine `color?.toLowerCase() === defectType.color` and `dirty` to achieve the expected results. 

allowCloseOutside={!dirty || color?.toLowerCase() === defectType.color}

Comment (Vadim73i) [reply app/src/pages/inside/projectSettingsPageContainer/content/defectTypes/modals/addEditDefectTypeModal/addEditDefectTypeModal.jsx:183]: The bug was that `dirty` was not correct.
We don't change the color in the end, but dirty=true.
And `!true && color?.toLowerCase() === defectType.color` =false

**Review by AmsterGet — CHANGES_REQUESTED** [review]
The main reason that `dirty` not worked as expected that Redux form initial value is in `lowerCase`, but copied from text field is in `upperCase`.
So the best solution here is to `lowerCase` value while setting to Redux form.
Try [`parse`](https://redux-form.com/8.2.2/docs/api/field.md/#-code-parse-value-name-gt-parsedvalue-code-optional-) or [normalize](https://redux-form.com/8.2.2/docs/api/field.md/#-code-normalize-value-previousvalue-allvalues-previousallvalues-name-gt-nextvalue-code-optional-).

**Pull request merged**